cpp/include/copying.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef COPYING_HPP
+#define COPYING_HPP
+
+#include "cudf.h"
+#include "types.hpp"
+
+/** 
+ * @brief Scatters the rows of a set of source columns into a set
+ * of destination columns.
+ * 
+ * Scatters the rows of the source columns into the destination columns according
+ * to a scatter map such that row "i" from the source columns will be scattered 
+ * to row "scatter_map[i]" in the destination columns.
+ * 
+ * The datatypes between coresponding columns in the source and destination columns
+ * must be the same.
+ * 
+ * The number of elements in the scatter_map must equal the number of rows in the source
+ * columns. 
+ * 
+ * If any index in scatter_map is outside the range of [0, num rows in destination_columns),
+ * the result is undefined.
+ * 
+ * If the same index appears more than once in scatter_map, the result is undefined.
+ * 
+ * @Param[in] source_columns The columns whose rows will be scattered
+ * @Param[in] scatter_map An array that maps rows in the input columns
+ * to rows in the output columns.
+ * @Param[out] destination_columns A preallocated set of columns with a number of 
+ * rows equal in size to the maximum index contained in scatter_map
+ * 
+ * @Returns GDF_SUCCESS upon successful completion
+ */
+gdf_error gdf_scatter(cudf::table const * source_table,
```

Comment:
This is the only API in cuDF that I know of that refers to cuff::table. I've never understood why cuDF has something called a table when the Python API calls it a DataFrame, and the name of the library has DataFrame in it. should cudf::table be renamed?

Reply:
DataFrame is a highly Python specific term where as "table" is more generic. I don't think of libcudf as being exclusively a Python "DataFrame" library.

Reply:
Also, Arrow calls it a `table` too: https://arrow.apache.org/docs/cpp/classarrow_1_1_table.html

Reply:
- I don't think that DataFrame is Python specific. DataFrames were first introduced in the R language. Spark has DataFrames and Spark is not Python specific (it was originally Scala).
 - The name of our library is cuDF, which is short for CUDA DataFrame. 
 - I agree that table is a more general name than DataFrame, but it is also more overloaded.
 - I don't agree that "column" is a more general concept than "series".

I do think it would be nice, if we are going to redesign the libcudf API, to aim for consistency with the cuDF API, both in naming and in key abstractions.

Reply:
My two cents, table invokes more of a database feel when data frame makes it more specific to a data processing engine. I would prefer if we standardize on DF, but I’m open to a discussion.

Reply:
> * I don't think that DataFrame is Python specific. DataFrames were first introduced in the R language. Spark has DataFrames and Spark is not Python specific (it was originally Scala).
> * The name of our library is cuDF, which is short for CUDA DataFrame.
> * I agree that table is a more general name than DataFrame, but it is also more overloaded.
> * I don't agree that "column" is a more general concept than "series".
> 
> I do think it would be nice, if we are going to redesign the libcudf API, to aim for consistency with the cuDF API, both in naming and in key abstractions.

- In my mind, "DataFrame" carries more meaning than "table". It refers to a high-level abstraction where users have expectations about specific features the abstraction supports. In contrast, "table" is a bare-bones abstraction for a collection of columns without any exceptions on which features it supports. ( I understand that a SQL Table is counter to the argument I'm making, but I see that as an exception to the rule. The term "table" existed long before any programming language.) In short, a "DataFrame" can be implemented on top of a "table" but not vice-versa. 

- Yes, cuDF is short for CUDA DataFrame, but cuDF is the combination of the Python & C++ layers. A `cudf::table` is the minimalist abstraction we use in C++ on top of which we implement the Python DataFrame (like I said, a DF can be implemented on top of a `table`). 

- In the C++ layer, I believe it would be misleading to call it a "DataFrame" because that term has semantic baggage.  What we want is a term for a simple data structure that represents `n` columns without any implication of higher level features. I believe this is why Arrow refers to it as a "table" instead of a DataFrame.

Ultimately I think it will be a potential source of confusion to conflate a "C++ DataFrame" with a Python DataFrame. If I'm talking to you about a DataFrame, you may not know if I'm talking about the Python abstraction or the C++ data structure.  In contrast if we call it a `cudf::table` then there is no confusion. A DataFrame is implemented on top of a `cudf::table` and everyone knows what each other is talking about. 

I'm mainly just playing devil's advocate here (plus it's fun to argue semantics :) ).  I'm not a data scientist nor an end user of cuDF or libcudf, so if you guys think its best for end users to call it a `cudf::dataframe`, I'll support that (as I said before, that's what I named it in the first place).